File: packages/seo/src/meta.js
// 页面 meta 的 DOM 管理核心（纯浏览器端）。
//
// 三方标签协同约定（与后端 seo 模块、各 app index.html 的 <!-- seo:meta --> 标记块配套）：
// - index.html 静态兜底标签带 data-seo="static"；
// - 后端 HTML 注入（整块替换标记块）产出的标签带 data-seo="ssr"；
// - 本模块插入的标签带 data-seo="client"。
// 应用某类标签前，先把同类的 static/ssr 节点摘下暂存，组件卸载时原样放回，
// 保证「详情页 → 无自定义 meta 的页面」回退路径上静态兜底不丢失。
// <title> 不做增删（浏览器只认第一个），只赋值 document.title。

const CLIENT = 'client'
const ROUTE_ROBOTS = 'client-robots'

function metaNamed(name, content) {
  const el = document.createElement('meta')
  el.setAttribute('name', name)
  el.setAttribute('content', content)
  return el
}

function metaProperty(property, content) {
  const el = document.createElement('meta')
  el.setAttribute('property', property)
  el.setAttribute('content', content)
  return el
}

// 每类可管理标签：DOM 选择器 + 节点工厂。og:title/og:description 由 title/description 派生，
// og:url 由 canonical 派生（见 applyPageMeta），故这里不区分来源、只按标签种类管理。
const KINDS = {
  description: {
    selector: 'meta[name="description"]',
    create: (v) => metaNamed('description', v),
  },
  robots: {
    selector: `meta[name="robots"]:not([data-seo="${ROUTE_ROBOTS}"])`,
    create: (v) => metaNamed('robots', v),
  },
  canonical: {
    selector: 'link[rel="canonical"]',
    create: (v) => {
      const el = document.createElement('link')
      el.setAttribute('rel', 'canonical')
      el.setAttribute('href', v)
      return el
    },
  },
  'og:title': { selector: 'meta[property="og:title"]', create: (v) => metaProperty('og:title', v) },
  'og:description': {
    selector: 'meta[property="og:description"]',
    create: (v) => metaProperty('og:description', v),
  },
  'og:type': { selector: 'meta[property="og:type"]', create: (v) => metaProperty('og:type', v) },
  'og:image': { selector: 'meta[property="og:image"]', create: (v) => metaProperty('og:image', v) },
  'og:url': { selector: 'meta[property="og:url"]', create: (v) => metaProperty('og:url', v) },
  jsonld: {
    selector: 'script[type="application/ld+json"]',
    create: (v) => {
      const el = document.createElement('script')
      el.setAttribute('type', 'application/ld+json')
      // JSON-LD 是脚本内容：转义 < 防止数据中出现 </script> 逃逸
      el.textContent = JSON.stringify(v).replaceAll('<', '\\u003c')
      return el
    },
  },
}

// kind -> 被摘下的 static/ssr 原始节点（卸载时放回）。模块级单例：
// 同一时刻页面上至多一个详情视图在管理 meta，无需按组件区分归属。
const stashed = new Map()

function upsertKind(kind, value) {
  const def = KINDS[kind]
  // 首次接管该类标签：摘下 static/ssr 兜底节点并暂存
  if (!stashed.has(kind)) {
    const originals = [...document.head.querySelectorAll(def.selector)].filter(
      (el) => el.dataset.seo !== CLIENT,
    )
    originals.forEach((el) => el.remove())
    stashed.set(kind, originals)
  }
  document.head
    .querySelectorAll(`${def.selector}[data-seo="${CLIENT}"]`)
    .forEach((el) => el.remove())
  const el = def.create(value)
  el.dataset.seo = CLIENT
  document.head.appendChild(el)
}

/**
 * 应用一组页面 meta。只处理有值的字段，未提供的字段保留现有（static/ssr）标签，
 * 这样后端已注入的详情页在前端数据未就绪前不会出现 meta 真空。
 */
export function applyPageMeta(desc) {
  if (!desc) return
  const { title, description, canonical, ogType, ogImage, jsonLd, robots } = desc
  if (title) {
    document.title = title
    upsertKind('og:title', title)
  }
  if (description) {
    upsertKind('description', description)
    upsertKind('og:description', description)
  }
  if (canonical) {
    upsertKind('canonical', canonical)
    upsertKind('og:url', canonical)
  }
  if (ogType) upsertKind('og:type', ogType)
  if (ogImage) upsertKind('og:image', ogImage)
  if (jsonLd) upsertKind('jsonld', jsonLd)
  if (robots) upsertKind('robots', robots)
}

/** 移除本模块插入的 client 标签，并放回先前摘下的 static/ssr 兜底节点。不回滚 document.title（交给路由钩子）。 */
export function resetPageMeta() {
  document.head
    .querySelectorAll(`[data-seo="${CLIENT}"]`)
    .forEach((el) => el.remove())
  for (const originals of stashed.values()) {
    originals.forEach((el) => document.head.appendChild(el))
  }
  stashed.clear()
}

/** 路由级 noindex：独立标签、独立开关，不与 applyPageMeta 的 robots 互相干扰。 */
export function setRouteNoindex(on) {
  const existing = document.head.querySelector(`meta[name="robots"][data-seo="${ROUTE_ROBOTS}"]`)
  if (on && !existing) {
    const el = metaNamed('robots', 'noindex')
    el.dataset.seo = ROUTE_ROBOTS
    document.head.appendChild(el)
  } else if (!on && existing) {
    existing.remove()
  }
}

File: backend/crates/server/src/modules/seo/meta.rs
//! PageMeta：一个内容页要注入的元数据，渲染为 head 标记块 HTML。
//!
//! 渲染产物整块替换 index.html 中的 `<!-- seo:meta --> … <!-- /seo:meta -->`；
//! 所有标签带 `data-seo="ssr"`，前端 @haruhi/seo 在数据就绪后按该属性接管替换。

use super::esc;

pub struct PageMeta {
    /// 完整标题（含站名后缀）。404 壳也要给默认标题——标记块被整体替换，title 不能缺。
    pub title: String,
    pub description: Option<String>,
    /// 规范 URL（绝对地址）；同时派生 og:url。
    pub canonical: Option<String>,
    /// og:type：article / book / product / website。
    pub og_type: &'static str,
    /// og:image（绝对 URL）。
    pub og_image: Option<String>,
    pub json_ld: Option<serde_json::Value>,
    /// Some("noindex")：内容不存在/不可见的 404 壳。
    pub robots: Option<&'static str>,
}

impl PageMeta {
    pub fn new(title: impl Into<String>) -> Self {
        Self {
            title: title.into(),
            description: None,
            canonical: None,
            og_type: "website",
            og_image: None,
            json_ld: None,
            robots: None,
        }
    }

    /// 内容不存在/不可见时的 404 壳 meta：默认标题 + noindex。
    pub fn not_found(default_title: &str) -> Self {
        Self {
            robots: Some("noindex"),
            ..Self::new(default_title)
        }
    }
}

fn tag_meta(out: &mut String, attr: &str, key: &str, content: &str) {
    out.push_str(&format!(
        "    <meta {attr}=\"{key}\" content=\"{}\" data-seo=\"ssr\" />\n",
        esc(content)
    ));
}

/// 渲染标记块（含起止注释本身）。
pub fn render_head_block(m: &PageMeta) -> String {
    let mut out = String::from("<!-- seo:meta -->\n");
    out.push_str(&format!("    <title>{}</title>\n", esc(&m.title)));
    tag_meta(&mut out, "property", "og:site_name", "凉宫春日应援团");
    tag_meta(&mut out, "property", "og:type", m.og_type);
    tag_meta(&mut out, "property", "og:title", &m.title);
    if let Some(d) = &m.description {
        tag_meta(&mut out, "name", "description", d);
        tag_meta(&mut out, "property", "og:description", d);
    }
    if let Some(c) = &m.canonical {
        out.push_str(&format!(
            "    <link rel=\"canonical\" href=\"{}\" data-seo=\"ssr\" />\n",
            esc(c)
        ));
        tag_meta(&mut out, "property", "og:url", c);
    }
    if let Some(i) = &m.og_image {
        tag_meta(&mut out, "property", "og:image", i);
    }
    if let Some(r) = m.robots {
        tag_meta(&mut out, "name", "robots", r);
    }
    if let Some(j) = &m.json_ld {
        // JSON-LD 在 <script> 内：转义 < 防数据中出现 </script> 逃逸
        let json = j.to_string().replace('<', "\\u003c");
        out.push_str(&format!(
            "    <script type=\"application/ld+json\" data-seo=\"ssr\">{json}</script>\n"
        ));
    }
    out.push_str("    <!-- /seo:meta -->");
    out
}

/// og:image 等相对地址绝对化：http(s) 原样；`/` 开头拼站点基址；
/// 其余视为 uploads 相对路径（如 fiction/covers/x.webp）。
pub fn absolutize(base: &str, url: &str) -> String {
    if url.starts_with("http://") || url.starts_with("https://") {
        url.to_string()
    } else if let Some(rest) = url.strip_prefix('/') {
        format!("{base}/{rest}")
    } else {
        format!("{base}/uploads/{url}")
    }
}

/// 富文本 → 摘要：剥标签、解常见实体、折叠空白、按字符数截断。
/// 输入是 ammonia 白名单净化过的 content_html，无需完整 HTML 解析器。
pub fn excerpt(html: &str, max_chars: usize) -> String {
    let mut text = String::with_capacity(html.len());
    let mut in_tag = false;
    for c in html.chars() {
        match c {
            '<' => in_tag = true,
            '>' => in_tag = false,
            _ if !in_tag => text.push(c),
            _ => {}
        }
    }
    let text = text
        .replace("&nbsp;", " ")
        .replace("&amp;", "&")
        .replace("&lt;", "<")
        .replace("&gt;", ">")
        .replace("&quot;", "\"")
        .replace("&#39;", "'");
    truncate_chars(
        &text.split_whitespace().collect::<Vec<_>>().join(" "),
        max_chars,
    )
}

/// 按字符数截断（多字节安全），超长加省略号。
pub fn truncate_chars(s: &str, max_chars: usize) -> String {
    if s.chars().count() <= max_chars {
        s.to_string()
    } else {
        let mut out: String = s.chars().take(max_chars.saturating_sub(1)).collect();
        out.push('…');
        out
    }
}

/// URL 路径段 percent-encode（uid 等自由文本进 canonical/og:url 前用）：
/// 保留 RFC 3986 unreserved 字符，其余按 UTF-8 字节编码。
pub fn encode_path_segment(s: &str) -> String {
    let mut out = String::with_capacity(s.len());
    for b in s.as_bytes() {
        match b {
            b'A'..=b'Z' | b'a'..=b'z' | b'0'..=b'9' | b'-' | b'.' | b'_' | b'~' => {
                out.push(*b as char)
            }
            _ => out.push_str(&format!("%{b:02X}")),
        }
    }
    out
}
